File: 826.cpp
struct Solution {
    int maxProfitAssignment(const vector<int>& difficulty, const vector<int>& profit, vector<int>& worker) {
        sort(worker.begin(), worker.end());
        int ptr = 0, n = profit.size(), maxP = 0, res = 0;
        vector<int> idx(n);
        iota(idx.begin(), idx.end(), 0);
        sort(idx.begin(), idx.end(), [&difficulty](const int &a, const int &b) {
            return difficulty[a] < difficulty[b];
        });
        for (int &w : worker){
            while (ptr < n && difficulty[idx[ptr]] <= w){
                maxP = max(maxP, profit[idx[ptr++]]);
            }
            res += maxP;
        }
        return res;
    }
};

File: sol/826/826.py
class Solution:
    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        worker.sort()
        jobs = sorted(zip(difficulty, profit))
        ptr = 0
        n = len(jobs)
        res = 0
        maxProfit = 0
        for w in worker:
            while ptr < n and jobs[ptr][0] <= w:
                maxProfit = max(maxProfit, jobs[ptr][1])
                ptr += 1
            res += maxProfit
        return res
